Support BigInt values

paul-marechal opened the pull request:
JSON doesn't constrain numeric values to be specific sizes. In most
languages this isn't an issue thanks to proper support for a wide
variety of number precision.

Unfortunately JavaScript struggles with integer numbers bigger than
`Number.MAX_SAFE_INTEGER`. This is due to JavaScript's implementation of
integer values as `double`. For anything bigger than this max safe
integer value JavaScript provides a `BigInt` type.

The first issue with this behavior is that when parsing JSON, there is
no way of telling which "number flavor" to expect, APIs just don't allow
us to define so. Another issue is that the default JSON
serializer/deserializer doesn't even know how to handle `BigInt`.

This commit fixes those issues as followed:

1. Use a custom JSON parser `json-bigint` that can serialize and
   deserialze `BigInt` values. A quirk is that it will only deserialize
   a number into a `BigInt` if the serialized value ends up being bigger
   than JavaScript's max safe integer. This means there is ambiguity
   when deserialing data.
2. To address the ambiguity issue, I added a TypeScript mapped type
   `Deserialized<T>` that given a type `T` will replace all fields
   that are either `bigint` or `number`. Note that this is only static
   typing to teach TypeScript about this behavior. Then when receiving
   messages, one has to define `Normalizer` functions which will take
   this ambiguous type as input and return a "normalized" version that
   matches the original type `T`. See this as post-processing to make
   sure the received data is using the proper data types in the right
   places.
3. To help with the normalization I added a `createNormalizer` helper
   function that should statically figure out what fields need to be
   normalized for a given type and ensure that those are properly
   handled.
4. Rewrite all the tests to validate this logic using test data as
   coming out of the current Trace Server.

paul-marechal commented:
Regarding 3) I wish I wouldn't have had to rewrite everything, but as far as I could tell nothing was being tested...

Huge thanks to @PatrickTasse for providing the test data.

Another thing is that implementing the parsing/normalization revealed some TSP messages to be incorrectly defined? If you look at the changes made to the TypeScript interfaces I had to add an optional marker (`?`) for a bunch of fields. There seems to be a mismatch between the TSP specification and what the Trace Server returns? If I had to guess, I'd say you guys need to do a pass on the fields again and properly specify what's required and what's optional.

paul-marechal commented:
Lastly this work is based on https://github.com/theia-ide/tsp-typescript-client/pull/28 so there might be a few common changes.

paul-marechal commented on src/protocol/serialization.ts:
```diff
@@ -0,0 +1,69 @@
+    : T
+    ;
+
+// export type MustBeNormalized<T> =
+//     T extends bigint | number
+//     ? T
+//     : T extends { [key: string]: any }
+//     ? {
+//         [K in keyof T as T[K] extends MustBeNormalized<T[K]> ? (
+//             K
+//         ) : never
+//     ]: MustBeNormalized<T[K]> }
+//     : never
+//     ;
+
+// type a = MustBeNormalized<{
+//     a: string
+//     b: number
+//     c: {
+//         d: string
+//         e: bigint
+//     }
+//     f: {
+//         g: string
+//     }
+// }>;
+
+// export type NormalizerDescriptor<T> =
+//     T extends bigint
+//     ? (number: bigint | number) => bigint
+//     : T extends number
+//     ? (number: bigint | number) => number
+//     : T extends { [key: string]: any }
+//     ? { [K in keyof T]: NormalizerDescriptor<T[K]> }
+//     : T extends (infer U)[]
+//     ? (array: U[]) => U[]
+//     : never
+//     ;
```
Left overs, disregard those for now. I wanted to make it so normalizers could be written as:

```ts
interface SomeType {
    ok1: string
    ok2: string[]
    notOk1: number
    notOk2: bigint
}
const SomeType = createNormalizer<SomeType>({
    notOk1: Number,
    notOk2: BigInt,
});
```

But the typings to achieve and enforce that are a bit tricky to come up with...

PatrickTasse commented on src/protocol/rest-client.ts:
```diff
@@ -1,77 +1,119 @@
-    }
 
+    public static get<T>(url: string, parameters?: Map<string, string>): Promise<TspClientResponse<BigintOrNumber<T>>>;
+    public static get<T>(url: string, parameters?: Map<string, string>, normalizer?: Normalizer<T>): Promise<TspClientResponse<T>>;
```
These don't seem necessary?

paul-marechal replied:
I think I actually need those to have TypeScript understand the two regimes we have.

See https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads

The first signature tells TypeScript that if we don't pass a normalizer, then we must get a `TspClientResponse<BigintOrNumber<T>>` because this is what the deserializer gives us.

The second signature informs TypeScript that if we pass a normalizer, then there's no need to worry about the `BigInt` issue and we can expect to get `TspClientResponse<T>`.

The third occurrence of the function is the implementation that must respect the two overloads.

The first signature is useful when `BigintOrNumber<T>` is strictly the same as `T`. This happens if the type doesn't contain any `number` or `bigint` fields. This behavior is relied upon in a couple places:

https://github.com/theia-ide/tsp-typescript-client/blob/1496d5ebfeb35f640fc528cf16e9df1b1e0b534f/src/protocol/tsp-client.ts#L197

https://github.com/theia-ide/tsp-typescript-client/blob/1496d5ebfeb35f640fc528cf16e9df1b1e0b534f/src/protocol/tsp-client.ts#L254

https://github.com/theia-ide/tsp-typescript-client/blob/1496d5ebfeb35f640fc528cf16e9df1b1e0b534f/src/protocol/tsp-client.ts#L275

https://github.com/theia-ide/tsp-typescript-client/blob/1496d5ebfeb35f640fc528cf16e9df1b1e0b534f/src/protocol/tsp-client.ts#L307

https://github.com/theia-ide/tsp-typescript-client/blob/1496d5ebfeb35f640fc528cf16e9df1b1e0b534f/src/protocol/tsp-client.ts#L355

https://github.com/theia-ide/tsp-typescript-client/blob/1496d5ebfeb35f640fc528cf16e9df1b1e0b534f/src/protocol/tsp-client.ts#L364

paul-marechal replied:
In your own words:

> It's required to be able to get a compilation error when a normalizer is missing.

PatrickTasse commented on src/protocol/rest-client.ts:
```diff
@@ -1,77 +1,119 @@
     }
 
+    public static post<T>(url: string, body?: any): Promise<TspClientResponse<BigintOrNumber<T>>>;
+    public static post<T>(url: string, body?: any, normalizer?: Normalizer<T>): Promise<TspClientResponse<T>>;
```
These don't seem necessary?

PatrickTasse commented on src/protocol/rest-client.ts:
```diff
@@ -1,77 +1,119 @@
     }
 
+    public static put<T>(url: string, body?: any): Promise<TspClientResponse<BigintOrNumber<T>>>;
+    public static put<T>(url: string, body?: any, normalizer?: Normalizer<T>): Promise<TspClientResponse<T>>;
```
These don't seem necessary?

PatrickTasse commented on src/protocol/rest-client.ts:
```diff
@@ -1,77 +1,119 @@
     }
 
+    public static delete<T>(url: string, parameters?: Map<string, string>): Promise<TspClientResponse<BigintOrNumber<T>>>;
+    public static delete<T>(url: string, parameters?: Map<string, string>, normalizer?: Normalizer<T>): Promise<TspClientResponse<T>>;
```
These don't seem necessary?

PatrickTasse commented on src/protocol/rest-client.ts:
```diff
@@ -1,77 +1,119 @@
+    }
+
+    protected static performRequest<T>(method: string, url: string, body?: any): Promise<TspClientResponse<BigintOrNumber<T>>>;
+    protected static performRequest<T>(method: string, url: string, body: any, normalizer?: Normalizer<T>): Promise<TspClientResponse<T>>;
```
These don't seem necessary?

PatrickTasse commented on src/protocol/rest-client.ts:
```diff
@@ -1,77 +1,119 @@
+
+    protected static performRequest<T>(method: string, url: string, body?: any): Promise<TspClientResponse<BigintOrNumber<T>>>;
+    protected static performRequest<T>(method: string, url: string, body: any, normalizer?: Normalizer<T>): Promise<TspClientResponse<T>>;
+    protected static async performRequest(method, url, body?, normalizer?) {
```
However it requires to declare it as normalizer? : Normalizer<T> here.

PatrickTasse replied:
That's Normalizer open bracket T close bracket.
Way to go, github.

paul-marechal replied:
Use backticks `` ` `` to wrap code stuff and it will work :)

paul-marechal replied:
I improved the typings as suggested.

paul-marechal commented on src/models/xy.ts:
```diff
@@ -18,6 +28,20 @@ export interface XYModel {
     series: XYSeries[];
 }
 
+const XYSeries: Normalizer<XYSeries> = input => {
+    const { seriesId, tags, xValues, yValues, ...rest } = input;
+    const xySeries: XYSeries = {
+        ...rest,
+        seriesId: assertNumber(seriesId),
+        xValues: xValues.map(Number),
+        yValues: yValues.map(Number),
+    };
+    if (tags !== undefined) {
+        xySeries.tags = tags.map(assertNumber);
+    }
+    return xySeries;
+};
```
@PatrickTasse explained to me how it might be common for `x/yValues` to actually come as `BigInt` values...

The issue I have is that if that is the case, then shouldn't the `XYSeries` type use `bigint` instead of `number`?

Alternatively, if we keep `number` the current implementation will lossly convert into a regular `number`. Isn't it better to throw here instead and let callers handle the error by not displaying bogus data?

PatrickTasse replied:
The yValues should definitely not be bigint, they can be floating-point values which would crash as bigint.

The xValues right now are timestamps so can be bigint-size, however I'm not sure if we want to restrict the protocol to time for the x-axis or allow any kind of XY chart.

In the current XY chart implementation, the xValues values don't even matter, the only thing that matters is how many xValues there are in the array. I'm not sure how long this will be the case.

PatrickTasse commented on src/protocol/rest-client.ts:
```diff
@@ -82,4 +124,19 @@ export class RestClient {
+     * into a regular `number` then it will be deserialized as `BigInt`.
+     */
+    protected static jsonParse(text: string): any {
+        return JSONBig.parse(text);
```
While testing I find that the JSONBig library crashes when parsing an XY output response, that looks like a showstopper!

Uncaught (in promise) SyntaxError: Cannot convert 101.76568027641497 to a BigInt
    at BigInt (<anonymous>)
    at number (parse.js:215)
    at value (parse.js:397)
    at array (parse.js:314)
    at value (parse.js:391)
    at object (parse.js:367)
    at value (parse.js:389)
    at array (parse.js:314)
    at value (parse.js:391)
    at object (parse.js:367)

PatrickTasse replied:
Yep, you can make the unit test fail but putting 101.76568027641497 as one of the yValues in the fixture.

paul-marechal replied:
I'll dig deeper later to see if it's a low hanging fix in that library.

PatrickTasse replied:
https://github.com/sidorares/json-bigint/issues/49

PatrickTasse commented on src/protocol/tsp-client.test.ts:
```diff
@@ -1,107 +1,277 @@
+      for (const xValue of serie.xValues) {
+        expect(typeof xValue).toEqual('number');
+      }
+      for (const yValue of serie.yValues) {
```
I expected it to expect the yValues to also have length 3?

PatrickTasse commented on src/protocol/serialization.ts:
```diff
@@ -0,0 +1,153 @@
+ * be deserialized as `bigint` or as `number` depending on its size.
+ * `BigintOrNumber` is a mapped type that reflects that behavior.
+ */
+export type BigintOrNumber<T> =
```
I'm a bit hesitant with the name 'BigintOrNumber' because the resulting type can contain other things than numbers.

I was thinking about 'DeNormalized' or 'UnNormalized' or 'NotNormalized'. I saw 'MustBeNormalized' below, that would have been a good one. Or how about 'AmbiguousNumbers'?

Food for thought...

paul-marechal commented:
I just contributed the last remaining improvements to the typings for the `createNormalizer` function.

I also rebased on master.

paul-marechal commented:
I rebased and squashed all the commits. I know this is a big change but ultimately it does a lot of good with the test suite. What bothers me is how I had to implement a mini serialization framework to handle the big int values in structs on a per-case basis. The root of the issue is that there's no type reflection in TypeScript: the interfaces are erased from the runtime. So if we want runtime mechanics based on interfaces, we have to go the way I've done: explicitly create functions for each type, as a separate entity.

paul-marechal commented:
cc @bhufmann @PatrickTasse please tell me if anything bothers you with this PR.

bhufmann commented:
> cc @bhufmann @PatrickTasse please tell me if anything bothers you with this PR.

@PatrickTasse could you please continue with your review?

haoadoresorange commented on tsp-typescript-client/src/protocol/serialization.ts:
```diff
@@ -0,0 +1,202 @@
+/**
+ * `true` if `T` must be normalized, `false` otherwise.
+ */
+export type MustBeNormalized<T> =
```
My apologize if this is not right place, I was looking through the PR and wonder how does this type work ? I can't seem to understand the concept, especially with the `0` and `1` type being used.

paul-marechal replied:
It's a pain.

1. I want to check if `Deserialized<T>` is assignable to `T`. On paper, this should only happen when something like `number` gets converted into `bigint | number` in `Deserialized<T>`. If `T` is "safe" then `Deserialized<T>` will be assignable to `T` and everyone is happy.
2. Let's assume we use `any` in our type `T`: then `Deserialized<T>` will be compatible with `T`, but I wanted this to fail because technically a `number` could slide into it and get deserialized as `bigint | number`.
3. Only `any` and `unknown` seem to be assignable to `unknown`
4. `0` is not assignable to `1`
5. To fix (2) I replace occurrences of `any` by `0` on one side, and by `1` on the other side so that fields typed as `any` forcefully conflict with one another.

Doing this ensures that `any` fields get marked as "must be normalized".

haoadoresorange commented:
I see that this PR also use the custom parser for serialization, is that necessary ? I don't see the ambiguity when serializing bigint. The reason I'm asking is the custom parser might take perf hit compare to current implementation (default parser + regex)

Provide that it's necessary, does it call `toJSON` under the hood (I suppose not) and suffer from the same prob that I made the PR #41 for ?

paul-marechal commented:
@haoadoresorange 

1. `RegExp`s can be slow and while it might work, the whole "let's find and replace bigint strings in the serialized string and replace those with a literal number" keeps me up at night. I haven't run benchmarks, but I'd be curious to see what's the perf hit between the two for large TSP messages.
2. The third-party JSON serializer/deserializer is still called on non-bigint values IIUC, so it will call `.toJSON` there.

haoadoresorange commented:
> I haven't run benchmarks, but I'd be curious to see what's the perf hit between the two for large TSP messages

I ran a quick benchmark on `stringify` (default + regex vs json-bigint) [here](https://github.com/haoadoresorange/json-bigint-benchmark). TL;DR: There's perf hit when there's no BigInt in the object, and about the same with BigInt

> The third-party JSON serializer/deserializer still calls the native implementation on non-bigint values IIUC, so it should call .toJSON there.

So custom implementation `BigInt.prototype.toJSON` shouldn't affect it then ?

PatrickTasse reviewed (approved):
Thank you for this awesome contribution!

bhufmann reviewed (approved):
Thanks a lot. Looks good to me. I tested within the theia-trace-extension and it works well. Thanks also to enhance the unit tests.
